lib/ansible/modules/cloud/docker/docker_plugin.py

```diff
@@ -0,0 +1,297 @@
+        name = self.get_plugin_name()
+        try:
+            plugin = self.dclient.plugins.get(name)
+        except NotFound:
+            return None
+        except APIError as e:
+            self.client.fail(text_type(e))
+
+        return plugin
+
+    def has_different_config(self):
+        """
+        Return the list of differences between the current parameters and the existing volume.
+
+        :return: list of options that differ
+        """
+        differences = []
+        if self.parameters.plugin_options:
+            if not self.existing_plugin.settings:
+                differences.append('plugin_options')
+            else:
+                existing_options_list = self.existing_plugin.settings['Env']
+                existing_options = parse_options(existing_options_list)
+
+                for key, value in iteritems(self.parameters.plugin_options):
+                    if ((not existing_options.get(key) and value) or
+                            not value or
+                            value != existing_options[key]):
+                        differences.append('plugin_settings.%s' % key)
+
+        return differences
+
+    def install_plugin(self):
+        if not self.existing_plugin:
+            if not self.check_mode:
+                try:
+                    self.existing_plugin = self.dclient.plugins.install(self.parameters.plugin_name, self.parameters.plugin_alias)
+                except APIError as e:
+                    self.client.fail(text_type(e))
+
+            self.results['actions'].append("Installed plugin %s" % self.get_plugin_name())
+            self.results['changed'] = True
+
+    def remove_plugin(self):
+        if self.existing_plugin:
+            if not self.check_mode:
+                try:
+                    self.existing_plugin.remove()
+                except APIError as e:
+                    self.client.fail(text_type(e))
+
+            self.results['actions'].append("Removed plugin %s" % self.get_plugin_name())
+            self.results['changed'] = True
+
+    def update_plugin(self):
+        if not self.check_mode:
+            try:
+                if self.existing_plugin.enabled:
+                    self.existing_plugin.disable()
+                self.existing_plugin.configure(prepare_options(self.parameters.plugin_options))
+                self.existing_plugin.enable(1)
+            except APIError as e:
+                self.client.fail(text_type(e))
+
+        self.results['actions'].append("Updated plugin %s settings" % self.get_plugin_name())
+        self.results['changed'] = True
+
+    def present(self):
+        differences = []
+        if self.existing_plugin:
+            differences = self.has_different_config()
+
+        if self.existing_plugin:
+            self.update_plugin()
+        else:
+            self.install_plugin()
+
+        if self.diff or self.check_mode or self.parameters.debug:
+            self.results['diff'] = differences
```

Comment:
That's **not** how a module should return differences. Some docker_* modules have done this in the past (and some might still do that), but it's simply wrong (and won't show up when the user runs `ansible-playbook --diff`).